control/controlclient/direct.go

```diff
@@ -1190,6 +1209,58 @@ func sleepAsRequested(ctx context.Context, logf logger.Logf, timeoutReset chan<-
 	}
 }
 
+// getNoiseClient returns the noise client, creating one if one doesn't exist.
+func (c *Direct) getNoiseClient(ctx context.Context) (*noiseClient, error) {
+	c.mu.Lock()
+	su := c.serverURL
+	np := c.noiseClient
+	c.mu.Unlock()
+	if np != nil {
+		return np, nil
+	}
+	k, err := c.getMachinePrivKey()
+	if err != nil {
+		return nil, err
+	}
+
+	np, err = newNoiseClient(ctx, k, su)
+	if err != nil {
+		return nil, err
+	}
+	c.mu.Lock()
+	c.noiseClient = np
```

Comment:
at this point the noiseClient might've already been assigned by a concurrent call to getNoiseClient who won the race.

do you want to check that here and close the one you just made on Line 1226 and return the winner one?